File: lastStoneWeight.py
# You are given an array of integers stones where stones[i] is the weight of the ith stone.
# We are playing a game with the stones. On each turn, we choose the heaviest two stones and smash them together. 
# Suppose the heaviest two stones have weights x and y with x <= y. The result of this smash is:
# If x == y, both stones are destroyed, and
# If x != y, the stone of weight x is destroyed, and the stone of weight y has new weight y - x.
# At the end of the game, there is at most one stone left.
# Return the weight of the last remaining stone. If there are no stones left, return 0.

class Solution(object):
    def lastStoneWeight(self, stones):
        stones = [-s for s in stones]
        heapq.heapify(stones)

        while len(stones) > 1:
            heaviest = heapq.heappop(stones)
            secondHeaviest = heapq.heappop(stones)
            if secondHeaviest == heaviest:
                continue
            elif secondHeaviest != heaviest:
                newStone = heaviest - secondHeaviest
                heapq.heappush(stones, newStone)
        
        if len(stones) == 1:
            return -stones[0]
        else:
            return 0
# Explanation
# Create a maxHeap from stones by negating all stones in stones
# While the length of stones is greater than 1 do the following: (since we only want the last stone, or 0 stones)
# Pop the heaviest and second heaviest stones and assign them to variables (According to the rules given, these stones will always be popped)
# If the stones are equal, continue as they have already been popped
# If the secondHeaviest is less than the heaviest, create a new stone and push it to the heap
# Check if the length of stones is 1, if so return the last stone (negate it so it returns positive)
# Otherwise return 0

File: lastStoneWeight.c
// You are given an array of integers stones where stones[i] is the weight of the ith stone.
// We are playing a game with the stones. On each turn, we choose the heaviest two stones and smash them together. 
// Suppose the heaviest two stones have weights x and y with x <= y. The result of this smash is:
// If x == y, both stones are destroyed, and
// If x != y, the stone of weight x is destroyed, and the stone of weight y has new weight y - x.
// At the end of the game, there is at most one stone left.
// Return the weight of the last remaining stone. If there are no stones left, return 0.

int compare(const void* a, const void* b) {
    return (*(int*)a - *(int*)b);
}

int lastStoneWeight(int* stones, int stonesSize){
    for (int i = 0; i < stonesSize; i++){
        stones[i] = -stones[i];
    }

    qsort(stones, stonesSize, sizeof(int), compare);
    while(stonesSize > 1){
        int heaviest = stones[0];
        int secondHeaviest = stones[1];

        if (secondHeaviest == heaviest){
            stones[0] = 0;
            stones[1] = 0;
        }
        else{
            stones[0] = heaviest - secondHeaviest;
            stones[1] = 0;
        }
        qsort(stones, stonesSize, sizeof(int), compare);
        stonesSize--;
    }
    return -stones[0];
}

// Explanation
// Compare function to assist qsort
// Loop over stones and negate them to create a maxHeap
// Sort the maxHeap
// Loop while size is greater than 1
// Initialize two variables to the heaviest and secondHeaviest
// If they are equal, set both to 0
// If they are not, reassign stones[0] to heaviest - secondHeaviest and set secondHeaviest to 0
// Resort the maxHeap and reduce the size
// Return the last element
